core/storage/exploration/gae_models.py

```diff
@@ -515,3 +515,86 @@ def get_recently_published(cls, limit):
         ).order(
             -ExpSummaryModel.first_published_msec
         ).fetch(limit)
+
+
+class StateIDMappingModel(base_models.BaseModel):
+    """State ID model for Oppia explorations.
+
+    This model maps each exploration version's state to a unique id.
+    Note: use the state id only for derived data, but not for data that’s
+    regarded as the source of truth, as the rules for assigning state id may
+    change in future.
+
+    The key of each instance is a combination of exploration id and version.
+    """
+
+    # The exploration id whose states are mapped.
+    exploration_id = ndb.StringProperty(indexed=True, required=True)
+
+    # The version of the exploration.
+    exploration_version = ndb.IntegerProperty(required=True)
+
+    # A dict which maps each state name to a unique id.
+    state_names_to_ids = ndb.JsonProperty(required=True)
+
+    # Latest state id that has been assigned to any of the states in any of
+    # of the versions of given exploration. New state IDs should be assigned
+    # from this value + 1.
+    largest_state_id_used = ndb.IntegerProperty(required=True)
```

Comment:
Index this field, too. Might be useful for the future if we want to get stats about how often states change their identity.

Reply:
Done